packages/block-editor/src/components/block-alignment-visualizer/index.js

```diff
@@ -0,0 +1,391 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+
+/**
+ * WordPress dependencies
+ */
+import { getBlockSupport, hasBlockSupport } from '@wordpress/blocks';
+import { Popover } from '@wordpress/components';
+import { useMergeRefs, useRefEffect } from '@wordpress/compose';
+import { useSelect } from '@wordpress/data';
+import {
+	createPortal,
+	useContext,
+	useEffect,
+	useMemo,
+	useRef,
+	useState,
+} from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import { BlockList } from '../';
+import { useLayout, LayoutStyle } from '../block-list/layout';
+import { __unstableUseBlockElement as useBlockElement } from '../block-list/use-block-props/use-block-refs';
+import useAvailableAlignments from '../block-alignment-control/use-available-alignments';
+import { store as blockEditorStore } from '../../store';
+import { getValidAlignments } from '../../hooks/align';
+import { getDistanceToNearestEdge } from '../../utils/math';
+
+const highlightedZoneEdges = [ 'right' ];
+function detectNearestZone( point, zones ) {
+	let candidateZone;
+	let candidateDistance;
+
+	zones?.forEach( ( zone ) => {
+		const [ distance ] = getDistanceToNearestEdge(
+			point,
+			zone.node.getBoundingClientRect(),
+			highlightedZoneEdges
+		);
+
+		if ( ! candidateDistance || candidateDistance > distance ) {
+			candidateDistance = distance;
+			candidateZone = zone;
+		}
+	} );
+
+	return candidateZone;
+}
+
+export default function BlockAlignmentVisualizer( {
+	allowedAlignments,
+	blockListClientId,
+	focusedClientId,
+	dragPosition,
+} ) {
+	const layout = useLayout();
+	const { focusedBlockName, blockListBlockName } = useSelect(
+		( select ) => {
+			const { getBlockName } = select( blockEditorStore );
+
+			return {
+				focusedBlockName: getBlockName( focusedClientId ),
+				blockListBlockName: getBlockName( blockListClientId ),
+			};
+		},
+		[ focusedClientId, blockListClientId ]
+	);
+
+	const [ popoverAnchor, setPopoverAnchor ] = useState( null );
+	const [ coverElementStyle, setCoverElementStyle ] = useState( null );
+	const [ highlightedZone, setHighlightedZone ] = useState();
+	const zones = useRef( new Set() );
+	const focusedBlockElement = useBlockElement( focusedClientId );
+	const blockListBlockElement = useBlockElement( blockListClientId );
+	const rootBlockListElement = useContext(
+		BlockList.__unstableElementContext
+	);
+
+	useEffect( () => {
+		if ( dragPosition ) {
+			const nearestZone = detectNearestZone(
+				dragPosition,
+				zones.current
+			);
+			if ( nearestZone?.name !== highlightedZone ) {
+				setHighlightedZone( nearestZone?.name );
+			}
+		}
+	}, [ dragPosition ] );
```

Comment:
Could we rewrite this effect into a `useMemo`? See https://beta.reactjs.org/learn/you-might-not-need-an-effect#caching-expensive-calculations.

Reply:
Will take a look at that. I just pushed a commit that throttles `detectNearestZone`, and that should improve it to an extent.